numpy/core/src/multiarray/ctors.c

```diff
@@ -1802,7 +1808,17 @@ PyArray_GetArrayParamsFromObject(PyObject *op,
             return 0;
         }
         /* If object arrays are forced */
-        if (is_object) {
+        if (is_object != DISCOVERED_OK) {
+            if (is_object == DISCOVERED_RAGGED && requested_dtype == NULL) {
+                /* NumPy 1.18, 2019-11-01 */
+                if (DEPRECATE("Creating an ndarray from ragged nested sequences "
+                    "and automatic object dtype detection is deprecated. Use "
+                    "dtype=object or an exact dtype") < 0)
```

Comment:
Now that Eric pointed out that we may want to include "ragged", I am thinking we should actually bloat this up a lot more. Maybe even create a `VisibleDeprecationWarning` (although that is a bit annoying).

The reason is that we are currently aiming the error at users who want to keep the old behaviour, when actually most users who see it are likely people who simply mistyped and thus have a bug. For them "ragged" is probably not helpful except as a google term.

It is a bit long I know, but am considering something like:
> You are attempting to create an array, or doing an array operation, with uneven shape (ragged). This is not supported and may be an input error such as `np.array([[1], [1, 2]])` where the first list is shorter than the second. If you wish to create an array containing sequences of different length you must create an array with `dtype=object` in the future.

Reply:
Changed to a `VisibleDeprecationWarning`, extended the message, rebased off master, and fixed the deprecation version in e11d57a.

Although in my mind the difference between a VisibleDeprecationWarning and an Error is negligible: emitting a VisibleDeprecationWarning is so annoying it means there is no real deprecation period, users will, in practice, have to fix their code immediately.

Reply:
Searching for ["ragged nested sequences numpy"](https://duckduckgo.com/?q=ragged+nested+sequences+numpy&t=canonical&ia=web) the first hit is NEP 34, so I think this text is good enough.